### PR TITLE
fix: edge cases with flagd targeting

### PR DESCRIPTION
**Pull request by toddbaert**
* coerce booleans to stringified versions for matching
* implement new gherkin tests ([not yet released](https://github.com/open-feature/flagd-testbed/pull/85), will block PR for now)

### CODE REVIEW

File: providers/flagd/src/main/java/dev/openfeature/contrib/providers/flagd/resolver/process/InProcessResolver.java [hunk 1]
```diff
@@ -174,7 +174,7 @@ private <T> ProviderEvaluation<T> resolve(Class<T> type, String key,
                     resolvedVariant = flag.getDefaultVariant();
                     reason = Reason.DEFAULT.toString();
                 } else {
-                    resolvedVariant = jsonResolved;
+                    resolvedVariant = jsonResolved.toString(); // convert to string to support shorthand
```
Comment (toddbaert) [inline providers/flagd/src/main/java/dev/openfeature/contrib/providers/flagd/resolver/process/InProcessResolver.java:177]: This is the only functional change.

File: providers/flagd/src/main/java/dev/openfeature/contrib/providers/flagd/resolver/process/InProcessResolver.java [hunk 2]
```diff
@@ -198,7 +198,7 @@ private <T> ProviderEvaluation<T> resolve(Class<T> type, String key,
             // if this is a double and we are trying to resolve an integer, convert
             value = ((Double) value).intValue();
         }
-        if (!type.isAssignableFrom(value.getClass()) || !(resolvedVariant instanceof String)) {
+        if (!type.isAssignableFrom(value.getClass())) {
```
Comment (Kavindu-Dodan) [inline providers/flagd/src/main/java/dev/openfeature/contrib/providers/flagd/resolver/process/InProcessResolver.java:201]: What would happen when resolvedVarient is not a string. reason is at line 209, we have a casting to string. This could fail for non string casting exception 🤔 

I think we might need explicit handling to perform this

**Review by Kavindu-Dodan — APPROVED** [review]
I have [this](https://github.com/open-feature/java-sdk-contrib/pull/567/files#r1416188411) concern, otherwise looks fine

File: providers/flagd/src/main/java/dev/openfeature/contrib/providers/flagd/resolver/process/InProcessResolver.java (hunk 2 above)
Comment (toddbaert) [reply providers/flagd/src/main/java/dev/openfeature/contrib/providers/flagd/resolver/process/InProcessResolver.java:201]: I think you have a point.

We could expand the caught exception to include this. In that case, it'd still throw a `ParseError`, which I think is OK (you'd have incorrectly configured a flag in that case).

Comment (Kavindu-Dodan) [reply providers/flagd/src/main/java/dev/openfeature/contrib/providers/flagd/resolver/process/InProcessResolver.java:201]: [This](https://github.com/open-feature/java-sdk-contrib/blob/a2d319964f3ade5bfd6a05d587e0f403c9c09a87/providers/flagd/src/main/java/dev/openfeature/contrib/providers/flagd/resolver/process/InProcessResolver.java#L177-L178) logic which transforms variant to string solves my concern.

Along with this, I think we can do the following improvement 

 `final Object resolvedVariant` -> ` final String resolvedVariant`
 
 And avoid the cast. at line 209

Comment (Kavindu-Dodan) [reply providers/flagd/src/main/java/dev/openfeature/contrib/providers/flagd/resolver/process/InProcessResolver.java:201]: > you'd have incorrectly configured a flag in that case

This will fail at json parsing time. so we don't have to worry here. I think this resolved once we update the variable type

Comment (toddbaert) [reply providers/flagd/src/main/java/dev/openfeature/contrib/providers/flagd/resolver/process/InProcessResolver.java:201]: Excellent observation! I have made this change.

Comment (Kavindu-Dodan) [reply providers/flagd/src/main/java/dev/openfeature/contrib/providers/flagd/resolver/process/InProcessResolver.java:201]: Great 🤝
